README.md
# Action Recognition in Figure Ice Skating

> Next level figure skating ai analysis.

<p style="text-align:center; background-color: gray;"><img src="axel_paf.gif" width="200px"></p>

[Human Pose Recognition in Figure Ice Skating](#action-recognition-in-figure-ice-skating)

# TOC

- [Installation](#installation)
- [Start training process](#start-training-process)
- [Test skatingAI - extract background, predict body parts or keypoints](#test-skatingAI---extract-background,-predict-body-parts-or-keypoints)

---

In our research we investigated state-of-the art systems such as 
[OpenPose](https://github.com/CMU-Perceptual-Computing-Lab/openpose)
, [VideoPose3d](https://github.com/facebookresearch/VideoPose3D) or [wrnch.ai](https://wrnch.ai/) on 
their performance on figure ice skating. 
All systems we tested had problems with spins in figure ice skating. Especially the ones with 
elastic poses such as the Biellmann pirouette.

This is why we called this investigative project into life to understand, how pose recognition works in depth 
and how to increase performance of keypoint recognition in figure ice skating.  
With the here presented project we love to further spurn research work in the figure ice skating pose recognition or
general human pose estimation field of artistic sports, with a well structured, easy to understand and follow code project 
with Tensorflow 2.

Altogether, we created three sub-projects:

1) 🎓  exercises: 
    - here we followed the [Image segmentation](https://www.tensorflow.org/tutorials/images/segmentation) 
    tutorial from Tensorflow to understand key aspects of neural network creation with Tensorflow 2
2) 🖥️  skatingAI
    - the main Python module for our research code base
3) 🖋️  thesis
    - master thesis regarding this investigative research
    
    

# 🖥️  skatingAI

<p style="text-align:center; background-color: gray;"><img src="skatingAI/docs/img/alena_step_labeled2.png" width="200px"></p>

## Installation
We build our project upon [Tensorflow 2](https://www.tensorflow.org/) with Python 3.6. 
The easiest way to get started is to run everything inside a [Docker](https://docs.docker.com/) container.

1) Create `skating-ai` docker image from our Dockerfile
    ```bash
    docker build -t skating-ai skatingAI/Docker/
    ```
2) Run `skating-ai` docker container from the above created container, bind local file system into docker container 
and log into container *(you can choose a different container name, and restrict gpu usage for your likings)*
    ```bash
    docker run -it\
        --name skating-ai\
        --gpus all\
        --mount type=bind,source="$(pwd)"/skatingAI,target=/skatingAI\
        skating-ai:latest
    ```
> Prerequisite: you need an [nvidia-docker](https://github.com/NVIDIA/nvidia-docker) installation
>
3) Download video data and preprocess data from [3DPEOPLE DATASET](https://cv.iri.upc-csic.es/)
    ```bash
   cd skatingAI/Data/3dhuman
   python data_admin.py
   ```
   This download of data takes a couple of hours.


## Start training process
We created three modules for background extraction, body part prediction and keypoint detection.
These complete training modules can be found in `skatingAI/modules`.

For our ablation study we created several network architectures for the different networks in 
`skatingAI/nets`.

Start the training process inside the docker container:
```bash
python train.py
```
After starting the training process a menu will appear in which you can choose the paramaters for the network.
You can eiter choose the default ones, or define your custom parameters for gpu, batch_size etc.

The second menu which appears allows you to choose between which module you want to train background extraction, 
human part detection or keypoint detection.

The menu which appears after that allows you to choose a certain training setting, which we experimented with in our 
ablation study. Variations are between, network architecture, loss function and optimizer.
In `skatingAI/utils/hyper_paramater.py` you can define your own custom setting.

### Logging - images, weights, Tensorboard
If sticked to the default parameters, every 5 epochs the predicted images and weights will be logged to your local 
file system. So you can follow the training process.

Furthermore, several variables will be logged additionally to `skatingAI/logs` such as the accuracy or loss curves, 
histograms and distributions on how the network parameters progressed during training.
Additionally, we log some descriptive information about the training setting in the text information of the Tensorboard.


## Test skatingAI - extract background, predict body parts or keypoints
To test skatingAI on some random videos from the [3DPEOPLE DATASET](https://cv.iri.upc-csic.es/) run:
```bash
python predict.py
```
Or with the `--video` parameter you can specify the path to a custom video file:

```bash
python predict.py --video /path/to/file/video.avi
```








changelog.md
# Change Log

> In this work, an approach to detect human body joints with only little computation complexity will be researched with a special focus on figure ice skating. The resulting application should be able to run on mobile phones and detect body joints from just the camera image. 
>
> Therefore, the different implementations from OpenPose in C++, TensorFlow and PyTorch will be explored and inspected. These already include some experiments with different neuronal network architectures, however their keypoint detection amount varies as does their current work status or rather commit frequency. Furthermore, a new paper suggests ‘spatio temporal affinity fields’, which will be analyzed in this project.
>
> Different experiments and evaluations of the already existing research projects on keypoint detection will add to the final application. The training will be conducted on the provided DeepLearn server from the Stuttgart Media University. Therefore, the data from PoseTrack and the Coco dataset will be used.
>
> Since this research focuses on the detection of keypoints in figure ice skating, further considerations must be taken in mind: people in the background should not be detected, but only the person in focus for later analysis. The question there is, how to get the corespondent training data.

---


## 2018-04-01 [Status Report]: Improve Custom Training Loop, Evaluation, Loss Function
#### Learned Mask:
   ![Loss Calculation](skatingAI/evaluate/img/v7_1595_adam_1595_mask/0_predicted_mask_0_ouputs.png)  
#### Learned Classes
   ![Loss Calculation](skatingAI/evaluate/img/v7_1595_adam_1595_mask/155_output.png)  
- experiment with different net structures
    - try to train faster/ less memory
        - strided-down/-up convolution architecture: result converges worse
        - rescale in dataset generator: works far better (graphs will be included later)
    - u-net with mobilenetv2 as base (not able to learn body parts)
    - u-net as base for HPNet (converges worse than just HPNet and is slower)
    - improve HPNet with filter amount and layer amount according to evaluation analysis of layers
        - combine add and concat of layers
        - always add the initial Input layer to concat/ add
        - performs best
    ![HPNet Model Architecture](skatingAI/nets/imgs/hrnet_v8.png)  
- create new custom loss function for Class Imbalance
    - create weighed map according to a calculated graph with the relative body part distances
- reduce classes to 9
- add automatic dataset download and processing including saving in numpy compressed files

- <img src="https://latex.codecogs.com/svg.latex?\Large&space;\theta=y_t(x)-y_p(x);" />
- <img src="https://latex.codecogs.com/svg.latex?\Large&space;\delta = \theta * \mu[argmax(y_t)];" />  
- <img src="https://latex.codecogs.com/svg.latex?\Large&space;L = \sum_{i=0}^{n}\theta_i + \delta_i;" />

    
   ![Loss Calculation](skatingAI/docs/img/random/loss_calculation.png)  
    - the new loss increased training efficiency a lot (exact results will be included later)
    
 - improve evaluation
    - add functionality to test different net architectures 
    - add simple save functionality
    - show predicted images additionally to feature maps of layers
    - code refactoring & improvement
    - evaluation of best performing net HPNet with Adam optimizer after epoch 1595 [v7_1595_adam_1595_mask](skatingAI/evaluate/img/v7_1595_adam_1595_mask)
- evaluate different optimizers, learning rates & learning rate adjustments
#### Accuracy of different used optimizers
> red: adam; learning rate: 0.01  
> orange: nadam; learning rate: 0.01  
>blue: sgd  learning rate: 0.1  
>green: sgd with custom decay (increase on plateau; decrease with variable decay)  
>light blue: sgd with custom decay of 0.001  
>
**Accuracy:**
![accuracy_all](skatingAI/docs/img/v7/accuracy_all.svg)
**Relation of correct body part pixel to all body part pixel**
![accuracy_body_part_all](skatingAI/docs/img/v7/accuracy_body_part_all.svg)
**Comparison of loss**
![loss](skatingAI/docs/img/v7/loss_all.svg)
**Learning Rates of SGD and SGD with custom loss**
![lr](skatingAI/docs/img/v7/learning_rate.svg)
**Comparison of SGD and custom decay/ learning rates**
![loss](skatingAI/docs/img/v7/loss.svg)


## 2020-03-01 [Status Report]: Create Custom Training Loop (first step)

- Solve problems with Tensorflow 2:
    - tries to use available GPUs out-of-the box if configured correctly otherwise falls back to CPU
    - [official docker container](https://www.tensorflow.org/install/docker) from documentation uses Python 2 
    - try out multiple containers and find [tensorflow/tensorflow:latest-devel-gpu-py3](https://hub.docker.com/r/tensorflow/tensorflow/tags)
     to work with GPUs and Python 3 as expected
- create Python Generator function for loading the image data and not running out of memory
- cleanup code to use classes and type information
- create custom training loop
- integrate tensorboard logging information
- create HRNetwork with 3 building blocks for the body segmentation task: [Model Architecture](#custom-hrnet-model)

### Analysis initial test training

The [Prediction](#predition-after-last-epoch) after the 64th epoch shows compared to 
the [prediction](#predition-after-2nd-epoch), that the network learns to categorize the pixel data better to the different
classes after some epochs. At the beginning the prediction shows a variegated picture with multiple class predictions, 
with preceding epochs this prediction becomes clearer and less scattered.

However, the prediction at the end seems to target the torso as main class, 
probably because it's pixels appears most often in the picture. 
Therefore, the hyperparameters and loss functions must be further refined for getting decent training results.

This behavior is reflected by the loss and accuracy graphs.
The [accuracy graph](#accuracy-plot) starts to flatten after the 35th episode and reassembles 

skatingAI/run.py
import argparse
from collections import namedtuple
from datetime import datetime
from pathlib import Path

import numpy as np
import tensorflow as tf

from skatingAI.nets.hrnet.v7 import HRNet
from skatingAI.utils.DsGenerator import DsGenerator
from skatingAI.utils.losses import CILoss
from skatingAI.utils.utils import DisplayCallback, set_gpus, Metric, Logger


class MainLoop(object):
    """this is the class handling the main loop

     Args:
            GPU:
            NAME:
            W_COUNTER:
            LR_START:
            OPTIMIZER_DECAY:
            BATCH_SIZE:
            PREFETCH_BATCH_BUFFER:
            EPOCH_STEPS:
            EPOCHS:
            EPOCH_LOG_N:
     """

    def __init__(self, GPU: int, NAME: str, W_COUNTER: int, optimizer: str, LR_START: float, OPTIMIZER_DECAY: float,
                 BG: bool,
                 BATCH_SIZE=3,
                 PREFETCH_BATCH_BUFFER=1, EPOCH_STEPS=64, EPOCHS=5555, EPOCH_LOG_N=5, EPOCH_SGD_PLATEAUCHECK=50):
        self.BATCHSIZE: int = BATCH_SIZE
        self.PREFETCH_BATCH_BUFFER: int = PREFETCH_BATCH_BUFFER
        self.EPOCH_STEPS: int = EPOCH_STEPS
        self.EPOCHS: int = EPOCHS
        self.EPOCH_LOG_N: int = EPOCH_LOG_N
        self.EPOCH_SGD_PLATEAUCHECK: int = EPOCH_SGD_PLATEAUCHECK

        self.GPU: int = GPU
        self.NAME: str = NAME
        self.W_COUNTER: int = W_COUNTER
        self.LR_START: float = LR_START
        self.SGD_CLR_DECAY_RATE = [1e-5, 1e-4, 1e-3, 0.01]
        self.SGD_CLR_DECAY_COUNTER = 0
        self.OPTIMIZER_DECAY: float = OPTIMIZER_DECAY
        self.OPTIMIZER_NAME = optimizer
        self.step_custom_lr = 0

        self.N_CLASS: int = 9
        self.IMG_SHAPE: [int, int, int] = (240, 320, 3)

        set_gpus(GPU)

        self.iter, self.sample_frame, self.sample_mask = self._generate_dataset(BG)
        self.model = self._get_model()
        self.optimizer = self._get_optimizer()

        self.loss_fn = CILoss(self.N_CLASS)

    def _generate_dataset(self, BG: bool):
        self.generator = DsGenerator(resize_shape=(240, 320))

        sample_pair = next(self.generator.get_next_pair(BG))

        self.IMG_SHAPE = sample_pair['frame'].shape
        self.N_CLASS = np.max(sample_pair['mask']).astype(int) + 1

        train_ds = self.generator.build_iterator(self.IMG_SHAPE, self.BATCHSIZE, self.PREFETCH_BATCH_BUFFER)

        return train_ds.as_numpy_iterator(), sample_pair['frame'], sample_pair['mask']

    def _get_model(self) -> tf.keras.Model:
        hrnet = HRNet(self.IMG_SHAPE, int(self.N_CLASS))

        model = hrnet.model
        model.summary()

        if self.W_COUNTER != -1:
            model.load_weights(f"./ckpt{self.GPU}/hrnet-{self.W_COUNTER}.ckpt")
        else:
            self.W_COUNTER = 0

        tf.keras.utils.plot_model(
            model, to_file=f'nadins_{self.NAME}_e.png', show_shapes=True, expand_nested=True)
        tf.keras.utils.plot_model(
            model, to_file=f'nadins_{self.NAME}.png', show_shapes=True, expand_nested=False)

        return model

    def _get_optimizer(self):

        if self.OPTIMIZER_NAME == "adam":
            optimizer = tf.keras.optimizers.Adam(learning_rate=self.LR_START, epsilon=1e-8, amsgrad=True)  # 0.001
        elif self.OPTIMIZER_NAME == "nadam":
            optimizer = tf.keras.optimizers.Nadam(learning_rate=self.LR_START, beta_1=0.9, beta_2=0.999, epsilon=1e-07)
        elif self.OPTIMIZER_NAME == "sgd_clr":
            self.OPTIMIZER_DECAY = self.SGD_CLR_DECAY_RATE[self.SGD_CLR_DECAY_COUNTER]
            if self.SGD_CLR_DECAY_COUNTER + 1 < len(self.SGD_CLR_DECAY_RATE):
                self.SGD_CLR_DECAY_COUNTER += 1

            optimizer = tf.keras.optimizers.SGD(learning_rate=self.LR_START, momentum=0.9, decay=self.OPTIMIZER_DECAY,
                                                nesterov=True)
        else:
            optimizer = tf.keras.optimizers.SGD(learning_rate=self.LR_START, momentum=0.9, decay=self.OPTIMIZER_DECAY,
                                                nesterov=True)

        return optimizer

    def _track_logs(self, log_detail: bool = False):
        file_writer = tf.summary.create_file_writer(
            f"{Path.cwd()}/logs/metrics/{self.NAME}/{datetime.now().strftime('%Y_%m_%d__%H_%M')}")
        file_writer.set_as_default()
        progress_tracker = DisplayCallback(self.model, self.sample_frame, self.sample_mask, file_writer, self.EPOCHS,
                                           self.GPU)
        log1 = Logger(log=log_detail)
        log2 = Logger(log=True)

        return file_writer, progress_tracker, log1, log2

    def _calculate_metrics(self):
        self.metric_correct_px.append(self.loss_fn.correct_predictions.astype(np.float32))
        self.metric_correct_px_body_part.append(self.loss_fn.correct_body_part_pred.astype(np.float32))
        self.metric_acc_body_part.append(
            (self.loss_fn.correct_body_part_pred / self.loss_fn.body_part_px_n_true).astype(np.float32))
        self.metric_loss.append(self.loss_value)

    def start_train_loop(self):

        train_acc_metric = tf.keras.metrics.Accuracy()
        file_writer, progress_tracker, log_v, log2 = self._track_logs()

        self.metric_correct_px, self.metric_correct_px_body_part, self.metric_acc_body_part, self.metric_acc, self.metric_loss, self.metric_avg_acc_body_part = (
            Metric('correct_px'),
            Metric('correct_body_part_px'),
            Metric('accuracy_body_part'),
            Metric('accuracy'),
            Metric('loss'),
            Metric('avg_acc_body_part')
        )

        for epoch in range(self.W_COUNTER, self.EPOCHS):
            log2.log(message=f"Start of epoch {epoch}", block=True)

            for step in range(self.EPOCH_STEPS):
                log_v.log(message=f"Step {step}", block=True)
                batch = next(self.iter)
                log_v.log(message=f"got batch")

                with tf.GradientTape() as tape:
                    logits = self.model(batch['frame'], training=True)
                    self.loss_value = self.loss_fn(batch['mask'], logits)

                log_v.log(message=f"Calculated Logits & loss value")
                grads = tape.gradient(self.loss_value, self.model.trainable_weights)
                log_v.log(message=f"Calculated Grads")

                self.optimizer.apply_gradients(zip(grads, self.model.trainable_weights))
                log_v.log(message=f"Optimized gradients")

                max_logits = tf.argmax(logits, axis=-1)
                max_logits = max_logits[..., tf.newaxis]
                train_acc_metric(batch['mask'], max_logits)
                self._calculate_metrics()
                self.step_custom_lr += 1

            self.metric_avg_acc_body_part.append(self.metric_acc_body_part.get_median(False))
            if epoch > 0 and epoch % self.EPOCH_SGD_PLATEAUCHECK == 0 and self.OPTIMIZER_NAME == 'sgd_clr':
                if self.metric_avg_acc_body_part.is_curve_steep() == False:
                    self.optimizer = self._get_optimizer()
                    self.step_custom_lr = 0
                    log2.log('adjusted optimizer')

            if epoch % self.EPOCH_LOG_N == 0:
                if self.OPTIMIZER_NAME == 'sgd':
                    lr = self.LR_START * (
                            1. / (1. + self.OPTIMIZER_DECAY * (epoch - self.W_COUNTER) * self.EPOCH_STEPS))
                elif self.OPTIMIZER_NAME == 'sgd_clr':
                    lr = self.LR_START * 1 / (1 + self.OPTIMIZER_DECAY * self.step_custom_lr)
                else:
                    lr = self.LR_START
                log2.log(
                    message=f"learning_rate: [{lr}]    "
                            f"loss: [{self.loss_value.numpy()}]   "
                            f"acc_body_part: [{np.median(self.metric_acc_body_part.metrics)}]",
                    block=False)

                progress_tracker.on_epoch_end(epoch,
                                              loss=self.metric_loss.get_median(),
                                              metrics=[
                                                  self.metric_correct_px,
                                                  self.metric_correct_px_body_part,
                                                  self.metric_acc_body_part,
                                                  Metric(value=train_acc_metric.result(), name='accuracy'),
                                                  Metric(value=lr, name='learning_rate'),
                                              ],
                                              show_img=False)
                log2.log(message=f"Seen images: {self.generator.seen_samples}", block=True)
                train_acc_metric.reset_states()

        progress_tracker.on_train_end()
        log_v.log_end()


if __name__ == "__main__":
    ArgsNamespace = namedtuple('ArgNamespace',
                               ['gpu', 'name', 'wcounter', 'lr', 'decay', 'opt', 'bs', 'steps','epochs', 'log_n', 'bg'])

    parser = argparse.ArgumentParser(
        description='Train nadins awesome network :)')
    parser.add_argument('--gpu', default=1, help='Which gpu shoud I use?', type=int)
    parser.add_argument('--name', default="hrnet_v7", help='Name for training')
    parser.add_argument('--wcounter', default=-1, help='Weight counter', type=int)
    parser.add_argument('--lr', default=0.1, help='Initial learning rate', type=float)
    parser.add_argument('--decay', default=0.001, help='learning rate decay', type=float)
    parser.add_argument('--opt', default="sgd_clr", help='Optimizer [nadam, adam, sgd, sgd_clr]')
    parser.add_argument('--bs', default=3, help='Batch size', type=int)
    parser.add_argument('--steps', default=64, help='Epoch steps', type=int)
    parser.add_argument('--epochs', default=64, help='Epochs', type=int)
    parser.add_argument('--log_n', default=5, help='Epoch steps', type=int)
    parser.add_argument('--bg', default=True, help='Use training images with background', type=bool)
    args: ArgsNamespace = parser.parse_args()

    MainLoop(args.gpu, args.name, args.wcounter, args.opt, args.lr, args.decay, args.bg, BATCH_SIZE=args.bs,
             EPOCH_STEPS=args.steps,EPOCHS=args.epochs, EPOCH_LOG_N=args.log_n).start_train_loop()
